crates/runtime/src/datafusion/query/cache.rs

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{collections::HashSet, sync::Arc};
+
+use cache::{to_cached_record_batch_stream, CacheKey, QueryResultsCacheStatus, RawCacheKey};
+use datafusion::{
+    execution::{SendableRecordBatchStream, SessionState},
+    logical_expr::LogicalPlan,
+    physical_plan::memory::MemoryStream,
+    sql::TableReference,
+};
+use tracing::Span;
+
+use crate::{
+    datafusion::{error::find_datafusion_root, query::error_code::ErrorCode, DataFusion},
+    request::{CacheControl, CacheKeyType, RequestContext},
+};
+
+use super::{attach_query_tracker_to_stream, Query, QueryResult, QueryTracker};
+
+/// Returns `Plan` if the result is not cached and needs to be executed, otherwise returns `Cached`
+pub(super) enum PlanOrCached {
+    Plan(LogicalPlan, QueryTracker, RequestCacheManager),
+    Cached(QueryResult),
+}
+
+pub(super) struct RequestCacheManager {
+    pub(super) cache_status: QueryResultsCacheStatus,
+    pub(super) raw_cache_key: RawCacheKey,
+}
+
+impl RequestCacheManager {
+    fn new(cache_status: QueryResultsCacheStatus, raw_cache_key: RawCacheKey) -> Self {
+        Self {
+            cache_status,
+            raw_cache_key,
+        }
+    }
+
+    pub(super) fn should_cache_results(&self) -> bool {
+        !matches!(self.cache_status, QueryResultsCacheStatus::CacheDisabled)
+    }
+}
+
+enum CacheResult {
+    Hit(QueryResult),
+    MissOrSkipped(QueryTracker, QueryResultsCacheStatus),
+    WrongCacheKeyType(QueryTracker),
+}
+
+impl Query {
+    /// Returns a `LogicalPlan` if the result is not cached and needs to be executed, otherwise returns a cached `QueryResult`.
+    pub(super) async fn get_plan_or_cached(
+        df: &DataFusion,
+        session: &SessionState,
+        request_context: Arc<RequestContext>,
+        sql: &str,
+        tracker: QueryTracker,
+    ) -> super::Result<PlanOrCached> {
+        // Try to get cached results first from sql
+        let (tracker, cache_status) = match Self::try_get_cached_result(
```

Comment:
@phillipleblanc - could you please elaborate why we always try both CacheKey::String and CacheKey::LogicalPlan

Reply:
Only one would possible return results - having the check in `try_get_cached_result` makes it easier to follow the logic in this function.

Reply:
Got it 👍. Assume try_get_cached_result is cheap so we can call its underlying logic if we can technically avoid this.